Update auto-opt docs (#6037)
PyTorchLightning/pytorch-lightning: docs/source/common/lightning_module.rst

hparams
~~~~~~~

> After calling ``save_hyperparameters`` anything passed to init() is available via hparams.

Reviewer: ```suggestion
After calling ``save_hyperparameters`` anything passed to ``__init__()`` is available via the ``hparams`` attribute.
```

> After calling ``save_hyperparameters`` anything passed to ``__init__()`` is available via the ``hparams`` attribute.

Reviewer 2: It's not anything passed, right? @rohitgr7 

```suggestion
The arguments saved by calling ``save_hyperparameters`` passed through ``__init__()`` could be accessed by the ``hparams`` attribute.
```

Author: :+1: ... can you update this in your PR: https://github.com/PyTorchLightning/pytorch-lightning/pull/6056??

Reviewer 2: @Borda Updated it here https://github.com/PyTorchLightning/pytorch-lightning/pull/6065 :)

--------------

automatic_optimization
~~~~~~~~~~~~~~~~~~~~~~

> When set to False, Lightning does not automate the optimization process. This means you are responsible for your own

Reviewer: ```suggestion
When set to ``False``, Lightning does not automate the optimization process. This means you are responsible for your own
```

> When set to ``False``, Lightning does not automate the optimization process. This means you are responsible for your own
> optimizer behavior

Reviewer 3: ```suggestion
When set to ``False``, Lightning does not automate the optimization process. This means you are responsible for handling your optimizers. However, we do take care of precision and any accelerators used.
```

.. code-block:: python

    def __init__(self):

>         self.automatic_optimization = False

Reviewer 3: Same here

    def training_step(self, batch, batch_idx):
        # access your optimizers with use_pl_optimizer=False. Default is True
        opt = self.optimizers(use_pl_optimizer=True)

        loss = ...
        self.manual_backward(loss, opt)
        opt.step()
        opt.zero_grad()

This is not recommended when using a single optimizer, instead it's recommended when using 2+ optimizers

> AND you are an expert user. Most useful for research like RL, sparse coding and GAN research.

Reviewer: Spell it out?
```suggestion
AND you are an expert user. Most useful for research like reinforcement learning, sparse coding and GAN research.
```

> This is not recommended when using a single optimizer, instead it's recommended when using 2+ optimizers
> AND you are an expert user. Most useful for research like reinforcement learning, sparse coding and GAN research.

Reviewer 3: ```suggestion
This is recommended only if using 2+ optimizers AND if you know how to perform the optimization procedure properly. Note that automatic optimization can still be used with multiple optimizers by relying on the ``optimizer_idx`` parameter. Manual optimization is most useful for research topics like reinforcement learning, sparse coding, and GAN research.
```

> In the multi-optimizer case, ignore the optimizer_idx flag and use the optimizers directly

Reviewer: ```suggestion
In the multi-optimizer case, ignore the ``optimizer_idx`` argument and use the optimizers directly
```

.. code-block:: python

    def __init__(self):
        self.automatic_optimization = False

    def training_step(self, batch, batch_idx, optimizer_idx):
        # access your optimizers with use_pl_optimizer=False. Default is True
        (opt_a, opt_b) = self.optimizers(use_pl_optimizer=True)

        gen_loss = ...
        self.manual_backward(gen_loss, opt_a)
        opt_a.step()
        opt_a.zero_grad()

        disc_loss = ...
        self.manual_backward(disc_loss, opt_b)
        opt_b.step()
        opt_b.zero_grad()

--------------

example_input_array
~~~~~~~~~~~~~~~~~~~
Set and access example_input_array which is basically a single batch.

.. code-block:: python

    def __init__(self):
        self.example_input_array = ...
        self.generator = ...

    def on_train_epoch_end(...):
        # generate some images using the example_input_array
        gen_images = self.generator(self.example_input_array)

--------------

datamodule
~~~~~~~~~~
Set or access your datamodule.

.. code-block:: python

    def configure_optimizers(self):
        num_training_samples = len(self.datamodule.train_dataloader())
        ...

--------------

model_size
~~~~~~~~~~

> Get the model file size using ``self.model_size`` inside LightningModule.

Reviewer: what is the unit of this thing?
```suggestion
Get the model file size (in megabytes) using ``self.model_size`` inside LightningModule.
```

--------------

> automatic_optimization

Reviewer 4: I am confused. Shouldn't it be manual optimisation ?

Author: the property is called `automatic_optimization`

Reviewer 5: @tchaton ^^

~~~~~~~~~~~~~~~~~~~~~~
When set to False, Lightning does not automate the optimization process. This means you are responsible for your own
optimizer behavior

.. code-block:: python

    def __init__(self):
        self.automatic_optimization = False

    def training_step(self, batch, batch_idx):
        # access your optimizers with use_pl_optimizer=False. Default is True
        opt = self.optimizers(use_pl_optimizer=True)

        loss = ...
        self.manual_backward(loss, opt)
        opt.step()
        opt.zero_grad()

This is not recommended when using a single optimizer, instead it's recommended when using 2+ optimizers
AND you are an expert user. Most useful for research like RL, sparse coding and GAN research.

In the multi-optimizer case, ignore the optimizer_idx flag and use the optimizers directly

.. code-block:: python

    def __init__(self):
        self.automatic_optimization = False

>     def training_step(self, batch, batch_idx, optimizer_idx):

Reviewer 6: we should probably also make it so that the opt idx is not even part of the signature when doing manual opt, right? or perhaps it is already like this and just outdated docs.

Author: it's not like that in the code. Agreed opt_idx doesn't make sense in manual_optimization. will update.

        # access your optimizers with use_pl_optimizer=False. Default is True
        (opt_a, opt_b) = self.optimizers(use_pl_optimizer=True)

>         gen_loss = ...
>         self.manual_backward(gen_loss, opt_a)
>         opt_a.step()
>         opt_a.zero_grad()
> 
>         disc_loss = ...
>         self.manual_backward(disc_loss, opt_b)
>         opt_b.step()
>         opt_b.zero_grad()

Reviewer 4: ```suggestion
        gen_loss = ...
        opt_a.zero_grad()
        self.manual_backward(gen_loss, opt_a)
        opt_a.step()

        disc_loss = ...
        opt_b.zero_grad()
        self.manual_backward(disc_loss, opt_b)
        opt_b.step()
```

Author: you sure about this? I just checked a few tests related to this in there it's like:
```python
manual_backward()
step()
zero_grad()
```

Reviewer 3: We use zero_grad after step throught Lightning, however, recently we have noticed that GANs can fail to learn if it is not done before backward. see: #5545

Author: yep makes sense.. since there might be some left over gradients in the generator here from `disc_loss.backward()`

--------------

example_input_array
~~~~~~~~~~~~~~~~~~~
Set and access example_input_array which is basically a single batch.

.. code-block:: python

    def __init__(self):
        self.example_input_array = ...

>     def training_step(...):

Reviewer 4: That's a bit weird to have it in training_step. Might confused users.

Author: updated

        res = self(self.example_input_array)

--------------

> datamodule

Reviewer 4: I find this part really confusing. Looks like you are setting datamodule in LightningModule. I think it is bad practice.

Worse case, for configure_optimizers, `num_trianing_samples = self.trainer.datamodule.train_dataloader()`

Author: updated

~~~~~~~~~~
Set or access your datamodule.

.. code-block:: python

    def __init__(self):
        self.datamodule = LitDataModule(...)

    def configure_optimizers(self):
        train_len = len(self.datamodule.train_dataloader())

--------------

model_size
~~~~~~~~~~
Get the model file size.

.. code-block:: python

>     def training_step(...):

Reviewer 4: Not sure why model_size should be used in `training_step`.

Author: ok i'll just remove the example since I'm not sure about it's use-case.
